### PR TITLE
more robust code to fix #1933 and #2447

### LINKED ISSUE #1933: java.lang.IndexOutOfBoundsException  openfoodfacts.github.scrachx.openfood.views.adapters.EmbCodeAutoCompleteAdapter.a
java.lang.IndexOutOfBoundsException: 
  at java.util.ArrayList.get (ArrayList.java:437)
  at openfoodfacts.github.scrachx.openfood.views.adapters.EmbCodeAutoCompleteAdapter.a (EmbCodeAutoCompleteAdapter.java:2)
  at openfoodfacts.github.scrachx.openfood.views.adapters.EmbCodeAutoCompleteAdapter.getItem (EmbCodeAutoCompleteAdapter.java)
  at android.widget.ArrayAdapter.createViewFromResource (ArrayAdapter.java:441)
  at android.widget.ArrayAdapter.getView (ArrayAdapter.java:407)
  at android.widget.AbsListView.obtainView (AbsListView.java:3189)
  at android.widget.DropDownListView.obtainView (DropDownListView.java:323)
  at android.widget.ListView.makeAndAddView (ListView.java:2197)
  at android.widget.ListView.fillDown (ListView.java:824)
  at android.widget.ListView.fillFromTop (ListView.java:885)
  at android.widget.ListView.layoutChildren (ListView.java:1952)
  at android.widget.AbsListView.onLayout (AbsListView.java:2961)
  at android.view.View.layout (View.java:20841)
  at android.view.ViewGroup.layout (ViewGroup.java:6401)
  at android.widget.FrameLayout.layoutChildren (FrameLayout.java:323)
  at android.widget.FrameLayout.onLayout (FrameLayout.java:261)
  at android.view.View.layout (View.java:20841)
  at android.view.ViewGroup.layout (ViewGroup.java:6401)
  at android.widget.FrameLayout.layoutChildren (FrameLayout.java:323)
  at android.widget.FrameLayout.onLayout (FrameLayout.java:261)
  at android.view.View.layout (View.java:20841)
  at android.view.ViewGroup.layout (ViewGroup.java:6401)
  at android.view.ViewRootImpl.performLayout (ViewRootImpl.java:2948)
  at android.view.ViewRootImpl.performTraversals (ViewRootImpl.java:2635)
  at android.view.ViewRootImpl.doTraversal (ViewRootImpl.java:1779)
  at android.view.ViewRootImpl$TraversalRunnable.run (ViewRootImpl.java:7810)
  at android.view.Choreographer$CallbackRecord.run (Choreographer.java:911)
  at android.view.Choreographer.doCallbacks (Choreographer.java:723)
  at android.view.Choreographer.doFrame (Choreographer.java:658)
  at android.view.Choreographer$FrameDisplayEventReceiver.run (Choreographer.java:897)
  at android.os.Handler.handleCallback (Handler.java:789)
  at android.os.Handler.dispatchMessage (Handler.java:98)
  at android.os.Looper.loop (Looper.java:164)
  at android.app.ActivityThread.main (ActivityThread.java:6940)
  at java.lang.reflect.Method.invoke (Method.java)
  at com.android.internal.os.Zygote$MethodAndArgsCaller.run (Zygote.java:327)
  at com.android.internal.os.ZygoteInit.main (ZygoteInit.java:1374)

### LINKED ISSUE #2447: java.lang.NullPointerException:    at openfoodfacts.github.scrachx.openfood.fragments.AddProductIngredientsFragment
``` 
Caused by: java.lang.NullPointerException: 
  at openfoodfacts.github.scrachx.openfood.fragments.AddProductIngredientsFragment.a (AddProductIngredientsFragment.java:70)
  at openfoodfacts.github.scrachx.openfood.views.AddProductActivity.c (AddProductActivity.java:2)
  at openfoodfacts.github.scrachx.openfood.fragments.AddProductOverviewFragment$3.a (AddProductOverviewFragment.java:83)
  at openfoodfacts.github.scrachx.openfood.fragments.AddProductOverviewFragment$3.a (AddProductOverviewFragment.java:2)
  at io.reactivex.internal.operators.single.SingleObserveOn$ObserveOnSingleObserver.run (SingleObserveOn.java:14)
  at io.reactivex.android.schedulers.HandlerScheduler$ScheduledRunnable.run (HandlerScheduler.java:2)
  at android.os.Handler.handleCallback (Handler.java:808)
  at android.os.Handler.dispatchMessage (Handler.java:101)
  at android.os.Looper.loop (Looper.java:166)
  at android.app.ActivityThread.main (ActivityThread.java:7529)
  at java.lang.reflect.Method.invoke (Method.java)
  at com.android.internal.os.Zygote$MethodAndArgsCaller.run (Zygote.java:245)
  at com.android.internal.os.ZygoteInit.main (ZygoteInit.java:921)
```

### PR DESCRIPTION
**Pull request by deniger**
- Fix #1933 more robust code to avoid the IndexOutOfBoundsException exception
- Fix #2447 avoid NullPointerException if the user tries to add an ingredient photo just after a scan
- Remove duplicated code

## Description

Describe the changes made and why they were made instead of how they were made.

## Related issues and discussion
#fixes #1933, #2447 
 
 Test done with Galaxy S8 Android 8
Product: https://world.openfoodfacts.org/cgi/product.pl?type=edit&code=8019428000037

### CODE REVIEW

**Comment by teolemon** [comment]
small conflict to solve

**Comment by teolemon** [comment]
@deniger

File: app/src/main/java/openfoodfacts/github/scrachx/openfood/fragments/BaseFragment.java [hunk 1]
```diff
@@ -61,4 +63,12 @@ public void refreshView(State state) {
             swipeRefreshLayout.setRefreshing(false);
         }
     }
+
+    protected State getStateFromActivityIntent(){
+        Intent intent = getActivity().getIntent();
+        if(intent!=null && intent.getExtras()!=null && intent.getExtras().getSerializable("state")!=null) {
```
[the remark this answers is not in the document]
Comment (deniger) [reply app/src/main/java/openfoodfacts/github/scrachx/openfood/fragments/BaseFragment.java]: ok will do this morning :)

[the remark this answers is not in the document]
Comment (deniger) [reply app/src/main/java/openfoodfacts/github/scrachx/openfood/fragments/BaseFragment.java]: @brandenfung  should be ok now ;)

**Comment by deniger** [comment]
Could we merge this one ?

**Comment by teolemon** [comment]
@deniger merging

**Pull request merged**